### PR TITLE
feat(SqlLab): Change Save Dataset Button to Split Save Query Button IV

### PR DESCRIPTION
**Pull request by lyndsiWilliams**
### SUMMARY

This changes the current save query button into a split save button if the database has "Allow this database to be explored" enabled. When the user clicks the caret section of the button, they can now save the query as a dataset.

`Bonus`: In order to create this button, I conveniently needed to fix the styling on the "Run" query split button. The split line now goes all the way through the button and the caret is centered.

#### BEFORE:
saveButtonBefore
runButtonBefore

#### AFTER:
splitSaveButtonAfter
saveButtonAfter
runButtonAfter

### TESTING INSTRUCTIONS

- From SqlLab, open a query for a database that has "Allow this database to be explored" enabled.
  - Observe the new split save button
    - Click the "Save" section of the button to see the save query modal
    - Click the caret `v` section of the button, then click "Save dataset" to see the save dataset modal
    - Both of these modals should be fully functional
- From SqlLab, open a query for a database that does NOT have "Allow this database to be explored" enabled.
  - Observe that the save button is a normal button, not split
    - Clicking this button should show the save query modal

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/SqlLab/components/ResultSet/index.tsx [hunk 1]
```diff
@@ -275,11 +250,19 @@ export default class ResultSet extends React.PureComponent<
               this.props.database?.allows_virtual_table_explore && (
                 <ExploreResultsButton
                   database={this.props.database}
-                  onClick={() => this.setState({ showSaveDatasetModal: true })}
```
Comment (hughhhh) [inline superset-frontend/src/SqlLab/components/ResultSet/index.tsx:278]: add back this line

File: superset-frontend/src/SqlLab/components/ResultSet/index.tsx [hunk 2]
```diff
@@ -275,11 +250,19 @@ export default class ResultSet extends React.PureComponent<
...
+                    // if (qid) {
+                    //   // This will open explore using the query as datasource
+                    //   window.location.href = `/explore/?dataset_type=query&dataset_id=${qid}`;
+                    // } else {
+                    //   this.setState({ showSaveDatasetModal: true });
+                    // }
+                    this.setState({ showSaveDatasetModal: true });
+                  }}
                 />
-                // In order to use the new workflow for a query powered chart, replace the
```
Comment (hughhhh) [inline superset-frontend/src/SqlLab/components/ResultSet/index.tsx:280]: add back these lines

File: superset-frontend/src/SqlLab/components/SaveDatasetModal/SaveDatasetModal.test.tsx [hunk 3]
```diff
@@ -55,9 +63,732 @@ const mockedProps = {
   datasource: testQuery,
 };
 
-describe('SaveDatasetModal RTL', () => {
-  it('renders a "Save as new" field', () => {
-    render(<SaveDatasetModal {...mockedProps} />, { useRedux: true });
+const testDatasetList = {
```
Comment (hughhhh) [inline superset-frontend/src/SqlLab/components/SaveDatasetModal/SaveDatasetModal.test.tsx:66]: can we put this in a separate fixture file?

File: superset-frontend/src/components/Chart/Chart.jsx [hunk 4]
```diff
@@ -296,27 +306,39 @@ class Chart extends React.PureComponent {
...
+                source={this.props.dashboardId ? 'dashboard' : 'explore'}
+                data-test={this.props.vizType}
+              />
+            </div>
+            {isLoading && !isDeactivatedViz && <Loading />}
+          </Styles>
+        </ErrorBoundary>
+        {showSaveDatasetModal && (
+          <SaveDatasetModal
+            key={Math.random()}
```
Comment (hughhhh) [inline superset-frontend/src/components/Chart/Chart.jsx:333]: why are we doing a `Math.random()` here for the key? couldn't we just set it as specific value?

File: superset-frontend/src/components/Chart/Chart.jsx [hunk 5]
```diff
@@ -253,6 +262,7 @@ class Chart extends React.PureComponent {
       width,
     } = this.props;
 
+    const { showSaveDatasetModal } = this.state;
```
Comment (AAfghahi) [inline superset-frontend/src/components/Chart/Chart.jsx:265]: same as above

File: superset-frontend/src/components/Chart/ChartErrorMessage.tsx [hunk 6]
```diff
@@ -42,6 +42,5 @@ export const ChartErrorMessage: React.FC<Props> = ({
     ...error,
     extra: { ...error.extra, owners },
   };
-
```
Comment (AAfghahi) [inline superset-frontend/src/components/Chart/ChartErrorMessage.tsx:45]: nit: add this space back?

File: superset-frontend/src/SqlLab/components/ResultSet/index.tsx (hunk 1 above)
Comment (lyndsiWilliams) [reply superset-frontend/src/SqlLab/components/ResultSet/index.tsx:278]: Added back in [`this commit`](https://github.com/apache/superset/pull/20852/commits/61bdc42398bc3e8f458d1ba27536a4322b8e5097).

File: superset-frontend/src/SqlLab/components/ResultSet/index.tsx (hunk 2 above)
Comment (lyndsiWilliams) [reply superset-frontend/src/SqlLab/components/ResultSet/index.tsx:280]: Added back in [`this commit`](https://github.com/apache/superset/pull/20852/commits/61bdc42398bc3e8f458d1ba27536a4322b8e5097).

File: superset-frontend/src/SqlLab/components/ResultSet/index.tsx [hunk 7]
```diff
@@ -275,11 +250,19 @@ export default class ResultSet extends React.PureComponent<
...
+                    // and other times it will not.  We need this attribute consistently for this to work
+                    // const qid = this.props?.query?.results?.query_id;
+                    // if (qid) {
+                    //   // This will open explore using the query as datasource
+                    //   window.location.href = `/explore/?dataset_type=query&dataset_id=${qid}`;
+                    // } else {
+                    //   this.setState({ showSaveDatasetModal: true });
+                    // }
+                    this.setState({ showSaveDatasetModal: true });
+                  }}
```
[the remark this answers is not in the document]
Comment (lyndsiWilliams) [reply superset-frontend/src/SqlLab/components/ResultSet/index.tsx]: Removed in [`this commit`](https://github.com/apache/superset/pull/20852/commits/61bdc42398bc3e8f458d1ba27536a4322b8e5097).

File: superset-frontend/src/SqlLab/components/SaveDatasetModal/SaveDatasetModal.test.tsx (hunk 3 above)
Comment (lyndsiWilliams) [reply superset-frontend/src/SqlLab/components/SaveDatasetModal/SaveDatasetModal.test.tsx:66]: Yeah, once I'm done fixing this test file I'll be cleaning it up.

File: superset-frontend/src/components/Chart/Chart.jsx (hunk 4 above)
Comment (lyndsiWilliams) [reply superset-frontend/src/components/Chart/Chart.jsx]: This was removed in [`this commit`](https://github.com/apache/superset/pull/20852/commits/61bdc42398bc3e8f458d1ba27536a4322b8e5097).

File: superset-frontend/src/SqlLab/components/SaveQuery/index.tsx [hunk 8]
```diff
@@ -118,9 +129,7 @@ export default function SaveQuery({
     setDescription(e.target.value);
   };
 
-  const toggleSave = () => {
-    setShowSave(!showSave);
-  };
+  const toggleSave = () => setShowSave(!showSave);
```
[the remark this answers is not in the document]
Comment (lyndsiWilliams) [reply superset-frontend/src/SqlLab/components/SaveQuery/index.tsx]: Good call, I cleaned this up in [`this commit`](https://github.com/apache/superset/pull/20852/commits/61bdc42398bc3e8f458d1ba27536a4322b8e5097).

File: superset-frontend/src/SqlLab/components/ResultSet/index.tsx [hunk 9]
```diff
@@ -29,19 +29,14 @@ import {
   SaveDatasetModal,
 } from 'src/SqlLab/components/SaveDatasetModal';
 import { UserWithPermissionsAndRoles } from 'src/types/bootstrapTypes';
-import { EXPLORE_CHART_DEFAULT } from 'src/SqlLab/types';
```
[the remark this answers is not in the document]
Comment (lyndsiWilliams) [reply superset-frontend/src/SqlLab/components/ResultSet/index.tsx]: Yeah there seems to be some weird rebase stuff on this PR, thanks for catching all these! Re-added in [`this commit`](https://github.com/apache/superset/pull/20852/commits/61bdc42398bc3e8f458d1ba27536a4322b8e5097).

File: superset-frontend/src/SqlLab/components/ResultSet/index.tsx [hunk 10]
```diff
@@ -218,26 +213,6 @@ export default class ResultSet extends React.PureComponent<
     }
   }
 
-  createExploreResultsOnClick = async () => {
```
[the remark this answers is not in the document]
Comment (lyndsiWilliams) [reply superset-frontend/src/SqlLab/components/ResultSet/index.tsx]: Re-added in [`this commit`](https://github.com/apache/superset/pull/20852/commits/61bdc42398bc3e8f458d1ba27536a4322b8e5097).

File: superset-frontend/src/components/Chart/Chart.jsx [hunk 11]
```diff
@@ -25,6 +25,7 @@ import { PLACEHOLDER_DATASOURCE } from 'src/dashboard/constants';
 import Loading from 'src/components/Loading';
 import { EmptyStateBig } from 'src/components/EmptyState';
 import ErrorBoundary from 'src/components/ErrorBoundary';
+import { SaveDatasetModal } from 'src/SqlLab/components/SaveDatasetModal';
```
[the remark this answers is not in the document]
Comment (lyndsiWilliams) [reply superset-frontend/src/components/Chart/Chart.jsx]: Removed in [`this commit`](https://github.com/apache/superset/pull/20852/commits/61bdc42398bc3e8f458d1ba27536a4322b8e5097).

File: superset-frontend/src/components/Chart/Chart.jsx [hunk 12]
```diff
@@ -122,8 +123,10 @@ const MonospaceDiv = styled.div`
 class Chart extends React.PureComponent {
   constructor(props) {
     super(props);
+    this.state = { showSaveDatasetModal: false };
```
[the remark this answers is not in the document]
Comment (lyndsiWilliams) [reply superset-frontend/src/components/Chart/Chart.jsx]: Removed in [`this commit`](https://github.com/apache/superset/pull/20852/commits/61bdc42398bc3e8f458d1ba27536a4322b8e5097).

File: superset-frontend/src/components/Chart/Chart.jsx [hunk 13]
```diff
@@ -136,6 +139,12 @@ class Chart extends React.PureComponent {
     }
   }
 
+  toggleSaveDatasetModal = () => {
```
[the remark this answers is not in the document]
Comment (lyndsiWilliams) [reply superset-frontend/src/components/Chart/Chart.jsx]: Removed in [`this commit`](https://github.com/apache/superset/pull/20852/commits/61bdc42398bc3e8f458d1ba27536a4322b8e5097).

File: superset-frontend/src/components/Chart/Chart.jsx (hunk 5 above)
Comment (lyndsiWilliams) [reply superset-frontend/src/components/Chart/Chart.jsx]: Removed in [`this commit`](https://github.com/apache/superset/pull/20852/commits/61bdc42398bc3e8f458d1ba27536a4322b8e5097).

File: superset-frontend/src/components/Chart/ChartErrorMessage.tsx (hunk 6 above)
Comment (lyndsiWilliams) [reply superset-frontend/src/components/Chart/ChartErrorMessage.tsx]: Weird rebase stuff haha! Added back in [`this commit`](https://github.com/apache/superset/pull/20852/commits/61bdc42398bc3e8f458d1ba27536a4322b8e5097).

**Review by AAfghahi — APPROVED** [review]
LGTM

**Review by hughhhh — APPROVED** [review]

**Pull request merged**